### DISCUSSION TITLE
iPLAssの構成について

### DISCUSSION THREAD

**Discussion by fujita0804** [dentsusoken/iPLAss#850] [Q&A] [answered]
初めまして、藤田と申します。
Lowcodeのアプリケーションをいくつか検討しておりまして、こちらのiPLAssに興味をもっております。
ローカル環境で試験的に使用させて頂いているのですが、表題の件が気になっておりまして、質問させて頂きました。

以下、質問内容となります。
iPLAssの構成としては、標準の画面表示用のHTMLやCSS等はDBのメタデータを取得して表示させている形になっているのでしょうか？
加えて、EQLを使用している理由としては上記のメタデータを利用する上で必要だからでしょうか？

お忙しい所申し訳ありませんが、回答頂ければ幸いです。
以上となります。

**Comment by HiguchiKiyoshi** [comment, marked as answer]
@fujita0804 
iPLAssにご関心を寄せていただき、ありがとうございます。

ご質問いただきました件に関しましては、藤田さまご推察の通りとなります。

補足いたしますと、
標準で構成される画面につきましては、メタデータ（Entity定義、EntityView定義等）をもとに実行時に動的に構成する形となります。
俗にいう、ソースコード生成型ではなく、実行エンジン型に近いとご理解いただけたらと存じます。

EQLに関しまして、もう少し具体的な例で申しますと、
Entityに定義したプロパティ名が単純にRDBのテーブルのカラム名と同名になる形ではない（汎用性、変更容易性等の考慮のため）ので、そのプロパティ名-カラム名のマッピング処理等を行う必要があったりします。
また、iPLAssではEQLで発行されるクエリーに対して、操作者の権限による参照範囲の限定などを自動で行う機能が実装されていたりします。
それらの処理を挟み込むために独自のEQLといった言語を定義しております。

**Reply by fujita0804** [reply]
@HiguchiKiyoshi 
早々のご回答ありがとうございます。
非常に助かりました。

オープンソースという事もあり、標準の画面周りをカスタマイズしてみようかと思っていたのですが、メタデータベースで生成されるのであれば、少し難しいですね。

これに関しては完全に私の興味の範囲なので申し訳ないのですが、iPLAssにて利用されている「EQL」というのはoracle社がリファレンスを出している「EQL」とは完全に別の物になるのでしょうか？
以下、oracle社リファレンス
https://docs.oracle.com/cd/E53235_01/server.761/es_eql/src/ceql_intro_sql_compare.html

**Reply by HiguchiKiyoshi** [reply]
@fujita0804 
標準画面のカスタマイズに関しましては、拡張のためのフックポイントを設けておりまして、その範囲で可能なことであれば、iPLAss本体のソースコードを直接修正せずとも行っていただくことは可能となっております。

例えば、以下のような形です。
- CSSレベルで見た目を少し変更したい
->[標準のCSSをカスタマイズする](https://iplass.org/docs/developerguide/multitenant/index.html#_%E7%94%BB%E9%9D%A2%E8%A1%A8%E7%A4%BA%E8%A8%AD%E5%AE%9Agem%E3%81%AE%E8%A1%A8%E7%A4%BA%E8%A8%AD%E5%AE%9A)

- 詳細画面上に、追加のコンポーネント・コンテンツを表示したい
->[詳細画面に独自のhtmlタグやJavaScriptを組み込む](https://iplass.org/docs/developerguide/genericentitymanager/entityview/index.html#scriptsection)
->[独自の表示部品（※１）を作成し詳細画面に組み込む](https://iplass.org/docs/developerguide/genericentitymanager/entityview/index.html#templatesection)
※１ iPLAssにおいてはTemplateと称します。JSPもしくはGroovyTemplate形式で記述可能です。ロジックはサーバサイドで実行可能です

また、標準画面のメニューにカスタム画面へのリンクを設定することも可能ですので、まったく新しい機能・画面を作成いただき、それを組み込むことも可能となっております。
https://iplass.org/docs/sample/adminconsole/index.html#AdminConsole_Product_BulkUpdate



EQLに関しましては、iPLAss独自のものとなっております。

**Reply by fujita0804** [reply]
ご提示頂いた資料を拝見いたしました。
ありがとうございます。
styleseetやJavaScriptの指定について資料を確認させて頂いたんですが詳細が読み取れず、質問させて頂ければと思います。
指定方法については、ローカルで作成したものをPATH指定する形でいいんでしょうか？

**Reply by HiguchiKiyoshi** [reply]
はい、ご推察いただいた形となります。

例を示しますと、

1.  次の内容のファイルを `/styles/custom.css` のパスで配置（war内、もしくは開発環境上でのローカルファイルとして）。
```
h2 {
  font-weight: bold;
}
```
※styles配下のリソースにつきましては、iPLAss内で静的ファイルとして扱うようにデフォルトで設定されております。
※参考までに、CSS等のカスタムリソースをTemplateや静的コンテンツとして、adminConsoleからメタデータとして保存することも可能ではございます。話が複雑になってしまいますので、ここでの説明は割愛させていただきます。
https://iplass.org/docs/developerguide/customizing/index.html#Template
https://iplass.org/docs/developerguide/customizing/index.html#StaticResource

2.AdminConsoleのTenant設定画面にてテナントStyleSheetURLに `/styles/custom.css` を設定して保存。

上記にて、
標準のCSSに対して上記のCSSの設定が上書きされる形となり、検索画面、詳細画面の見出しがbold表示されるようになります。

**Reply by fujita0804** [reply]
丁寧なご回答、ありがとうございます。
その他設定については、もう少し触って確認をさせて頂こうと思います。

もし、iPLAssを利用する場合については、HPの問い合わせフォームからご連絡させて頂きます。

この度は、対応頂き誠にありがとうございました。

**Reply by fujita0804** [reply]
@HiguchiKiyoshi
申し訳ありません。
一点だけ確認させてください。
先ほどご教授頂いたCSSのURL指定の方法ですが、以下の場合はどのような指定方法になるでしょうか？
例
C/xxx/yyy/styles/custom.css
この場合はフルPATH指定になりますか？

**Reply by HiguchiKiyoshi** [reply]
URL指定に指定するパスはAPサーバ（Tomcat等）上からの公開パスである必要があります。
（正確に申しますと、デプロイされたアプリケーションのコンテキストパスからの相対パスを指定します）

APサーバにリソースを配置する一般的な方法はwar形式で配置することです。
もし、TomcatへiPLAssをインストーラwarを用いてデプロイいただいたとした場合、
`[Tomcatのインストールディレクトリ]/webapps/iplass/` といったディレクトリが作成されていると思われますので、
その配下に `/styles/custom.css` として配置いただければと思います。

ただし上記手順はあくまで簡易的な動作検証のための手順です。
iPLAssをバージョンアップ（warを更新）した場合、当該ファイルは削除されてしまいます。
実運用する場合は、
- war内にカスタムコンテンツを組み込む（通常のJavaでの開発同様warをビルドする）：Javaベースでの開発経験があることが望ましいです
- カスタムコンテンツをメタデータとして管理する（adminConsole経由で保存する）
といったような運用になろうかと思われます。

また、
今回例示させていただきましたカスタマイズ例が悪かったかもしれません。
申し訳ございません。。。
どの程度のカスタマイズ性を有するか説明するにあたり、

- 全体的に適用するためのフックポイント
- 個別ページに適用するためのフックポイント
- またフルスクラッチに近しい開発も可能

と柔軟性をご説明させていただこうと思いました。

CSSカスタム例は、iPLAssの標準画面全体的にカスタマイズを適用したい場合の例であり（またそのカスタムリソースの登録の仕方も複数の方法あり、Java開発者を想定した説明してしまったかもしれません）、
個別の特定ページにカスタム処理を入れたい場合は、
->[詳細画面に独自のhtmlタグやJavaScriptを組み込む](https://iplass.org/docs/developerguide/genericentitymanager/entityview/index.html#scriptsection)
の方法で直接styleを記述いただくことで事足りたかもしれません。。。

**Reply by fujita0804** [reply]
ご回答頂きありがとうございます！
当方、java開発の経験は少なく理解に至っておらず申し訳ありませんでした。。

> 個別の特定ページにカスタム処理を入れたい場合は、
> ->[詳細画面に独自のhtmlタグやJavaScriptを組み込む](https://iplass.org/docs/developerguide/genericentitymanager/entityview/index.html#scriptsection)
> の方法で直接styleを記述いただくことで事足りたかもしれません。。。
→こちらの目的としては初級者でもある程度開発を進められるツールを探しておりましたので、iPLAssの画面上で対応可能であれば、むしろ良いと考えております。

ご提示頂いた両方を私の方で試してみたいと思います。
